### DISCUSSION TITLE
How to extra all exif-metadata and strip some tags

### DISCUSSION THREAD

**Discussion by jishi92** [libvips/libvips#3167] [Q&A]
Now I use `vips_image_get_fields ` to get metadata but I find the result is diffent from the result from exiftool.

The input png:  

![exiftool_png](https://user-images.githubusercontent.com/22196414/202450997-e33b2ae9-73ac-49c6-8187-0b1f68fd2966.png)

All fields in libvips:
 `width height bands format coding interpretation xoffset yoffset xres yres filename vips-loader`

The exiftool result:   

```File Name                       : exiftool_png.png
Directory                       : .  
File Size                       : 332 bytes  
File Modification Date/Time     : 2022:06:23 11:43:37+08:00  
File Access Date/Time           : 2022:11:17 17:54:29+08:00  
File Inode Change Date/Time     : 2022:08:16 21:52:22+08:00  
File Permissions                : -rw-r--r--  
File Type                       : PNG  
File Type Extension             : png  
MIME Type                       : image/png  
Image Width                     : 15  
Image Height                    : 30  
Bit Depth                       : 8  
Color Type                      : RGB with Alpha  
Compression                     : Deflate/Inflate  
Filter                          : Adaptive  
Interlace                       : Noninterlaced  
Image Size                      : 15x30  
Megapixels                      : 0.000450
```

Then I don't know how to strip some tags such as 'Profile Description', 'creator'...

Can you give me some hints or code demo in any languge, preferably go or python.

**Comment by jcupitt** [comment]
Hi @jishi92,

The exiftool result is misleading -- those aren't real tags, it's just extra stuff about the file. `get_fields` is listing all the metadata that libvips tracks for that image.

**Reply by jishi92** [reply]
Oh, I think so.
I extra more tags in libvips by `vipsheader -a`.
I want to get all metadata about exif tags.
Which fun can do it ?

**Reply by jcupitt** [reply]
They should all be there. I see for example:

```
$ vipsheader -a k2.jpg
k2.jpg: 1450x2048 uchar, 3 bands, srgb, jpegload
width: 1450
height: 2048
bands: 3
format: uchar
coding: none
interpretation: srgb
xoffset: 0
yoffset: 0
xres: 2.835
yres: 2.835
filename: k2.jpg
vips-loader: jpegload
jpeg-multiscan: 0
jpeg-chroma-subsample: 4:2:0
exif-data: 186 bytes of binary data
resolution-unit: in
exif-ifd0-Orientation: 1 (Top-left, Short, 1 components, 2 bytes)
exif-ifd0-XResolution: 72009/1000 (72.009, Rational, 1 components, 8 bytes)
exif-ifd0-YResolution: 72009/1000 (72.009, Rational, 1 components, 8 bytes)
exif-ifd0-ResolutionUnit: 2 (Inch, Short, 1 components, 2 bytes)
exif-ifd0-YCbCrPositioning: 1 (Centred, Short, 1 components, 2 bytes)
exif-ifd2-ExifVersion: Exif Version 2.1 (Exif Version 2.1, Undefined, 4 components, 4 bytes)
exif-ifd2-ComponentsConfiguration: Y Cb Cr - (Y Cb Cr -, Undefined, 4 components, 4 bytes)
exif-ifd2-FlashpixVersion: FlashPix Version 1.0 (FlashPix Version 1.0, Undefined, 4 components, 4 bytes)
exif-ifd2-ColorSpace: 65535 (Uncalibrated, Short, 1 components, 2 bytes)
exif-ifd2-PixelXDimension: 1450 (1450, Long, 1 components, 4 bytes)
exif-ifd2-PixelYDimension: 2048 (2048, Long, 1 components, 4 bytes)
orientation: 1
```

Perhaps your libvips has been built without exif support?

**Reply by jishi92** [reply]
I don't think my first image is appropriate as an example.
I got exif data from another one.

**Reply by jcupitt** [reply]
Oh, do you mean exif in png? It's not very well supported, see eg.:

https://stackoverflow.com/questions/9542359/does-png-contain-exif-data-like-jpg

libvips doesn't yet support exif read and write, though I suppose we could.

Since libvips won't load or save exif from png, there's no need to strip it.

**Reply by jishi92** [reply]
Oh, an unlucky message.
I would continue to use exiftool.
Perhaps you can tell me how to read and write exif data in C.
I could try do it, if I have time.

**Reply by jishi92** [reply]
By the way, usually png images don't have exif data ?

**Reply by jcupitt** [reply]
Exif in PNG is pretty rare I think.

It wouldn't be hard to add to libvips -- in `spngload.c` you'd need to call `spng_get_exif()` to read out the exif data, then attach that to the vips image. Likewise, in `spngsave.c` you'd need to read any exif from the vips image and write to spng with `spng_set_exif()`. You can see the code to get and set exif on a vips image in the jpeg loader.

**Reply by jcupitt** [reply]
... I'll add it, I have a moment.

**Reply by jcupitt** [reply]
I made a PR, though it's not quite working. I'll fix it tomorrow.

https://github.com/libvips/libvips/pull/3168

**Reply by jcupitt** [reply]
... I finished the PR, it seems to work.

**Reply by jishi92** [reply]
That's really good news.  I am gratitude to you for helping me.
It looks work for png image .
Libvips could extra exif data from jpeg,png and heif? What about other type images like webp etc.
I don't know exactly what to do and I still want to a demo about my question title.

**Reply by jcupitt** [reply]
libvips already gets all the metadata from jpeg, tiff, webp, png and heif, I think (exif, ipct, xmp, icc, jfif, etc.). Do you have any examples of images where it misses some?

**Reply by jishi92** [reply]
I don't think I made myself clear.
I don't have any examples of images missing exif metadata.
I just don't know how to do in gpvips/pyvips .
I  don't see  examples in these projects.
And I hope you can give me some hints.

**Reply by jcupitt** [reply]
It's just `get_fields` in pyvips, I think you are already doing everything.

You can see all the metadata that libvips fetches from an image with `vipsheader -a` at the command-line, for example:

```
 $ vipsheader -a k2.jpg
k2.jpg: 1450x2048 uchar, 3 bands, srgb, jpegload
width: 1450
height: 2048
bands: 3
format: uchar
coding: none
interpretation: srgb
xoffset: 0
yoffset: 0
xres: 2.835
yres: 2.835
filename: k2.jpg
vips-loader: jpegload
jpeg-multiscan: 0
jpeg-chroma-subsample: 4:2:0
exif-data: 186 bytes of binary data
resolution-unit: in
exif-ifd0-Orientation: 1 (Top-left, Short, 1 components, 2 bytes)
exif-ifd0-XResolution: 72009/1000 (72.009, Rational, 1 components, 8 bytes)
...
```

And those are the field names that `get_fields` returns. 

I don't think there is any other useful metadata, but if you have an example of a file where useful metadata is missing, I'll certainly try to add support.

**Reply by jishi92** [reply]
I used `get_fields` and got all  field names(tags name).But I don't get the value of the corresponding field（tag）.
In your example:
`exif-ifd0-Orientation: 1 (Top-left, Short, 1 components, 2 bytes)
`
I can get `exif-ifd0-Orientation` but don't know how to get `1 (Top-left, Short, 1 components, 2 bytes)` in my code.

**Reply by jcupitt** [reply]
It should work. I see:

```
$ python3
Python 3.10.7 (main, Nov  2 2022, 18:49:29) [GCC 12.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import pyvips
>>> x = pyvips.Image.new_from_file("k2.jpg")
>>> x.get("exif-ifd0-Orientation")
'1 (Top-left, Short, 1 components, 2 bytes)'
>>> 
```

**Comment by jcupitt** [comment]
Looking at your original question again, `Profile description` and `Creator` sounds like metadata tags from the ICC profile, not from the image. You have to remove the ICC profile if you want to remove these tags.

**Reply by jishi92** [reply]
If I remove the ICC profile and I guess the color may be changed.

**Reply by jcupitt** [reply]
You could extract the profile, strip that in some other software, then reattach it.

**Reply by jishi92** [reply]
Yes, I do it using exiftool now. Extract the value from the tags I specified and then determine if the value needs to be cleared.
I'm just wondering if I can do it with libvips.

**Reply by jcupitt** [reply]
libvips will let you strip image metadata, but it doesn't support stripping metadata inside ICC profiles.

**Reply by jishi92** [reply]
Thanks, I got it.

**Reply by jishi92** [reply]
Hi @jcupitt :
I’m still wondering if you consider supporting strip metadata inside ICC profiles, like exiftool.
I think it could be add to do list.This function will help many people.

**Reply by jcupitt** [reply]
Sorry, l I think that's out of scope. I'd use lcms or argyll for colour profile editing.
